[ONNX] Update 'Functionalize' pass to support pre-decomp graph; Drop 'aten_graph' arg for 'DynamoExporter'

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99667

Summary
- Previously this was required by and entangled with `tracing_mode=symbolic` for `dynamic` tracing.
  That is resolved by #99555 and its follow ups.
- Later decomposition pass will do graph lowering, so this step is duplicated.
- Updated `Functionalization` to workaround https://github.com/pytorch/pytorch/issues/99774#issuecomment-1527949391

Todo
- Training vs eval in dynamo_export
  So we are effectively exporting all models in traning mode by
  default. But for the sake of this export we are only interested in eval mode.
  The question is, should we call `model.eval()` in `dynamo_export`?
  Tests with model containing batch norm fails 'functionalization' in training mode.
  We are explicitly calling `model.eval()` for these model for now.
- Merge decomp and functionalize pass. Both calls into `make_fx`.
  Merging potentially increases performance. However it is unclear
  if it will result in different behavior.

Fixes #99662. (For the functionalization issue. Still need missing op support.)

BowenBao committed:
[ONNX] Drop 'aten_graph' arg for 'DynamoExporter'

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Drop 'aten_graph' arg for 'DynamoExporter'"


Summary
- Previously this was required by `tracing_mode=symbolic` for `dynamic` tracing.
  That argument will be dropped by #99555.
- Later decomposition pass will do graph lowering, so this step is duplicated.
- Functionalization currently cannot work properly on aten level graph.
  So it must happen before lowering & decompositions.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Drop 'aten_graph' arg for 'DynamoExporter'"


Summary
- Previously this was required by `tracing_mode=symbolic` for `dynamic` tracing.
  That argument will be dropped by #99555.
- Later decomposition pass will do graph lowering, so this step is duplicated.
- Functionalization currently cannot work properly on aten level graph.
  So it must happen before lowering & decompositions.
- Introduce `ReplaceInplacePostFunctionalization` pass to replace inplace variant ops with outplace version.
  These ops are created by aten graph lowering and decomposition post functionalization. They
  won't be doing any real mutation as it is expected to have been handled by functionalization.

Workaround to unblock #99662.

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/dynamo_exporter.py:
```diff
@@ -164,10 +164,7 @@ def export(self) -> torch.onnx.ExportOutput:
         # TODO(wechi): There are several symbolic tracing mechanisms to convert
```
I think you can delete this as well, as we now only have one API.

titaiwangms commented on torch/onnx/_internal/fx/dynamo_exporter.py:
```diff
@@ -164,10 +164,7 @@ def export(self) -> torch.onnx.ExportOutput:
         # matured.
-        # TODO(titaiwang): Set `tracing_mode` according to `self.options.dynamic_shapes`
-        graph_module, graph_guard = torch._dynamo.export(
-            wrapped_model, *args, aten_graph=True, **kwargs
```
Curious that did aten_graph works in any bit when decomposition_table is not provided?

titaiwangms replied:
I suspect this is not even duplicated. It's just to enable symbolic mode.

BowenBao replied:
Commented below. It forces the `make_fx` to happen inside `dynamo.export`. And we had no choice because `dynamic=True` must require `tracing_mode='symbolic'`, which requires `aten_graph=True`. Not anymore.

titaiwangms commented on torch/onnx/_internal/fx/fx_exporter.py:
```diff
@@ -298,22 +298,32 @@ def export_fx_to_onnx(
-
         # ONNX does not support views and mutations.
         # Functionalize to get a semantically equivalent graph without mutations.
+        # NOTE: Functionalize must run before decomposition and aten graph lowering.
```
I guess I don't understand the difference between decomposition and aten graph lowering. Could you point me to the lines?

titaiwangms replied:
It seems to me that make_fx has nothing to do with aten_graph. Or aten_graph is part of decomposition table?

BowenBao replied:
In decomp pass https://github.com/pytorch/pytorch/blob/e9786149ab71874fad478109de173af6996f7eec/torch/onnx/_internal/fx/passes/decomp.py#L51-L56

In dynamo export
https://github.com/pytorch/pytorch/blob/e9786149ab71874fad478109de173af6996f7eec/torch/_dynamo/eval_frame.py#L829-L845

I'll update the comment. "aten graph lowering" is what make_fx does.

titaiwangms replied:
hmmm sounds like aten_graph == using decomposition_table? If that's the case, maybe we can consider stop mentioning aten graph, just say decompose?

BowenBao replied:
it's aten_graph == make_fx, and make_fx has the option to decompose with provided decomp_table. make_fx transforms graph of torch ops to graph of aten ops. If decomp_table is provided, make_fx decomposes along the way.

titaiwangms replied:
Thanks!

titaiwangms commented on torch/onnx/_internal/fx/fx_exporter.py:
```diff
@@ -298,22 +298,32 @@ def export_fx_to_onnx(
+        # These ops are created by aten graph lowering and decomposition post
+        # functionalization. No real mutation is expected as it should have been handled
+        # by functionalization.
+        module = passes.ReplaceInplacePostFunctionalization(module).run(*fx_module_args)
```
I think this deserves a TODO or FIXME given the fact we know it's a workaround.

BowenBao committed:
Update on "[ONNX] Drop 'aten_graph' arg for 'DynamoExporter'"


Summary
- Previously this was required by `tracing_mode=symbolic` for `dynamic` tracing.
  That argument will be dropped by #99555.
- Later decomposition pass will do graph lowering, so this step is duplicated.
- Functionalization currently cannot work properly on aten level graph.
  So it must happen before lowering & decompositions.
- Introduce `ReplaceInplacePostFunctionalization` pass to replace inplace variant ops with outplace version.
  These ops are created by aten graph lowering and decomposition post functionalization. They
  won't be doing any real mutation as it is expected to have been handled by functionalization.

Workaround to unblock #99662.

[ghstack-poisoned]

BowenBao committed:
Update functionalize pass. Note model train/eval mode topic on "[ONNX] Drop 'aten_graph' arg for 'DynamoExporter'"


Summary
- Previously this was required by `tracing_mode=symbolic` for `dynamic` tracing.
  That argument will be dropped by #99555.
- Later decomposition pass will do graph lowering, so this step is duplicated.
- Functionalization currently cannot work properly on aten level graph.
  So it must happen before lowering & decompositions.
- Introduce `ReplaceInplacePostFunctionalization` pass to replace inplace variant ops with outplace version.
  These ops are created by aten graph lowering and decomposition post functionalization. They
  won't be doing any real mutation as it is expected to have been handled by functionalization.

Workaround to unblock #99662.

[ghstack-poisoned]

BowenBao committed:
revert back functionalize and decomp order on "[ONNX] Drop 'aten_graph' arg for 'DynamoExporter'"


Summary
- Previously this was required by `tracing_mode=symbolic` for `dynamic` tracing.
  That argument will be dropped by #99555.
- Later decomposition pass will do graph lowering, so this step is duplicated.
- Functionalization currently cannot work properly on aten level graph.
  So it must happen before lowering & decompositions.
- Introduce `ReplaceInplacePostFunctionalization` pass to replace inplace variant ops with outplace version.
  These ops are created by aten graph lowering and decomposition post functionalization. They
  won't be doing any real mutation as it is expected to have been handled by functionalization.

Workaround to unblock #99662.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Drop 'aten_graph' arg for 'DynamoExporter'; Apply workaround in 'Functionalize' pass"


Summary
- Previously this was required by and entangled with `tracing_mode=symbolic` for `dynamic` tracing.
  That is resolved by #99555 and its follow ups.
- Later decomposition pass will do graph lowering, so this step is duplicated.
- Updated `Functionalization` to workaround https://github.com/pytorch/pytorch/issues/99774#issuecomment-1527949391

Todo
- Training vs eval in dynamo_export
  So we are effectively exporting all models in traning mode by
  default. But for the sake of this export we are only interested in eval mode.
  The question is, should we call `model.eval()` in `dynamo_export`?
  Tests with model containing batch norm fails 'functionalization' in training mode.
  We are explicitly calling `model.eval()` for these model for now.
- Merge decomp and functionalize pass. Both calls into `make_fx`.
  Merging potentially increases performance. However it is unclear
  if it will result in different behavior.

Workaround to unblock #99662.

[ghstack-poisoned]

BowenBao committed:
rebase on "[ONNX] Drop 'aten_graph' arg for 'DynamoExporter'; Apply workaround in 'Functionalize' pass"


Summary
- Previously this was required by and entangled with `tracing_mode=symbolic` for `dynamic` tracing.
  That is resolved by #99555 and its follow ups.
- Later decomposition pass will do graph lowering, so this step is duplicated.
- Updated `Functionalization` to workaround https://github.com/pytorch/pytorch/issues/99774#issuecomment-1527949391

Todo
- Training vs eval in dynamo_export
  So we are effectively exporting all models in traning mode by
  default. But for the sake of this export we are only interested in eval mode.
  The question is, should we call `model.eval()` in `dynamo_export`?
  Tests with model containing batch norm fails 'functionalization' in training mode.
  We are explicitly calling `model.eval()` for these model for now.
- Merge decomp and functionalize pass. Both calls into `make_fx`.
  Merging potentially increases performance. However it is unclear
  if it will result in different behavior.

Workaround to unblock #99662.

[ghstack-poisoned]

BowenBao committed:
rebase on "[ONNX] Drop 'aten_graph' arg for 'DynamoExporter'; Apply workaround in 'Functionalize' pass"


Summary
- Previously this was required by and entangled with `tracing_mode=symbolic` for `dynamic` tracing.
  That is resolved by #99555 and its follow ups.
- Later decomposition pass will do graph lowering, so this step is duplicated.
- Updated `Functionalization` to workaround https://github.com/pytorch/pytorch/issues/99774#issuecomment-1527949391

Todo
- Training vs eval in dynamo_export
  So we are effectively exporting all models in traning mode by
  default. But for the sake of this export we are only interested in eval mode.
  The question is, should we call `model.eval()` in `dynamo_export`?
  Tests with model containing batch norm fails 'functionalization' in training mode.
  We are explicitly calling `model.eval()` for these model for now.
- Merge decomp and functionalize pass. Both calls into `make_fx`.
  Merging potentially increases performance. However it is unclear
  if it will result in different behavior.

Workaround to unblock #99662.

[ghstack-poisoned]

titaiwangms commented:
Seems getting complicated. You might consider move Functionalization ahead of dropping aten_graph in the title.

BowenBao commented:
@pytorchbot merge

jon-chuang commented on torch/onnx/_internal/fx/passes/functionalization.py:
```diff
@@ -64,12 +69,37 @@ def __init__(
+            for inpt, input_functional in zip(flat_inputs, flat_inputs_functional):
+                if isinstance(input_functional, torch.Tensor):
+                    torch._sync(input_functional)
+                    inpt_new = torch._from_functional_tensor(input_functional)
```
Why is this inpt_new assigned?
